### PR TITLE
Allow calling `const unsafe fn` in `const fn` behind a feature gate

### PR DESCRIPTION
**Pull request by oli-obk**
cc #55607

r? @Centril

### CODE REVIEW

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr
Comment (Centril) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr]: There should be a help note here pointing towards the `min_const_unsafe_fn` gate.

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (Centril) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: It's not strictly necessary, but for good redundancy it would be nice to test specifically that `ptr::read` and `mem::transmute` error out.

**Review by Centril — COMMENTED** [review]
Tests look mostly good :)

**Comment by Centril** [comment]
Passing the review over to Ralf;

r? @RalfJung

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (RalfJung) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: So the body is not an implicit `unsafe` block? Love it :)

File: src/librustc/ty/constness.rs
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: What is this change about (and same some lines down)?

**Comment by RalfJung** [comment]
This doesn't actually allow any new operations in `unsafe const fn`, right? It just allows marking them as unsafe?

Unfortunately most of this is code that I am not at all familiar with, so I cannot judge the side-effects this may have. @Centril does that mean you would r+ and just want to get my opinion as well?

@oli-obk  I am a bit confused why there are two places where it makes the body of an `unsafe fn` not behave like an `unsafe` block: In `librustc_mir/build/mod.rs` and in `librustc_mir/transform/check_unsafety.rs`. Why is that?

File: src/librustc/ty/constness.rs
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: All calls to `is_min_const_fn` need to guarantee that the `DefId` is a a function. This way it dosn't need to, because the function checks it itself.

**Comment by oli-obk** [comment]
`check_unsafety` can actually `bug!` out for const fn where the body is treated as unsafe, I'll fix that.

**Comment by Centril** [comment]
@RalfJung 

> This doesn't actually allow any new operations in unsafe const fn, right? It just allows marking them as unsafe?

The intention is that calling `const unsafe fn`s should now be allowed in `unsafe { ... }` in `const unsafe? fn`.

> Unfortunately most of this is code that I am not at all familiar with, so I cannot judge the side-effects this may have. @Centril does that mean you would r+ and just want to get my opinion as well?

I mainly want someone who is more familiar with the compiler internals than me (e.g. you) to review the code and see that it is correct and such; but if you are not familiar with the code maybe let's pass it to a third person (e.g. eddyb) to also check it?

**Comment by oli-obk** [comment]
cc @arielb1 who wrote the original unsafety checker for MIR

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (RalfJung) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: Could we also have a test that does `&*x` on a raw ptr, i.e. "just" turns it into a ref? Just to make sure that is also caught.

I want to be certain that `from_bytes_with_nul_unchecked` would not get accepted by this. ;)

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (Centril) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: Entirely unrelated to this file...
But I think it might be good to replicate https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/min_const_fn/min_const_fn_libstd_stability.rs for `const unsafe fn` just to be on the safe side.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [inline src/librustc_mir/transform/check_unsafety.rs]: `MonConstFn` are the ones that are *allowed* in `min_const_fn`, right? So doesn't this code allow all of them?

The tests say it does not, but I do not understand why.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [inline src/librustc_mir/transform/check_unsafety.rs]: So first `MinConstFn` is a signal for what is allowed, and now it is used to get a hard error? This is confusing.

File: src/libsyntax/feature_gate.rs
Comment (RalfJung) [inline src/libsyntax/feature_gate.rs]: "and constructing layout-restricted types like `NonZero`"

**Comment by RalfJung** [comment]
> if you are not familiar with the code maybe let's pass it to a third person (e.g. eddyb) to also check it?

The unsafety checker I can at least read because it works on MIR (but I wouldn't call that familiarity); the `is_min_const_fn` etc. stuff and the HIR things I have no idea about at all.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: At this point it's not about allowing things anymore, but about how to report it. The same enum is used for both.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: This is how the other variants are used, too. There's no way around that except to start duplicating a lot of code.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [reply src/librustc_mir/transform/check_unsafety.rs]: Is that necessary? Can we have two enums, or at least not repurpose the same enum variant for antagonistic properties?

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [inline src/librustc_mir/transform/check_unsafety.rs]: There should be an entirely `const`-unrelated test making sure we catch things like the following:
```rust
#[rustc_layout_scalar_valid_range_start(1)]
#[repr(transparent)]
pub(crate) struct NonZero<T>(pub(crate) T);
fn main() {
    let _x = Some(NonZero(0));
}
```

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [reply src/librustc_mir/transform/check_unsafety.rs]: In which sense are they also used this way?

Why cannot we have `GeneralAndAllowedInMinConstFn` (strawman proposal :D ) as a new enum variant for, well, previously classified "general" violations that we do permit in `MinConstFn`?

Currently, how would you describe the interpretation of `UnsafetyViolationKind::MinContFn` if you had to write a doc comment for it? I think it'd be something like "either an operation outside of an `unsafe` block that requires `unsafe`, but that we *do* allow for min const; or an `unsafe` operation (maybe in an `unsafe` block) inside a min const fn that is not allowed there". These are two very different things. Which other things are similarly conflated in this check?

Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: I was down this rabbit hole yesterday and not able to find a clean solution.

I had multiple enum variants, that ended up with a bunch of `bug!` calls spread around.

I had separate enums, requiring loads of code duplication, because of closures/generators dumping a list of violations, means I'd need to convert that list back to a list of requirements. The unsafety checker as it is (and was before I ever touched it) used the violation struct both as requirements and violations.

Comment (RalfJung) [reply src/librustc_mir/transform/check_unsafety.rs]: Well I am afraid the way it works right now will take more time to review than I have this week-end. I'd have to disentangle this spaghetti mess first to understand how *anything* works.

Not good to have the *unsafety checker* out of all passes be so messy :(

Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: Note that this enum was created as a hack to support the future incompat lints for extern types and references to fields of packed structs.

This all will clean up nicely when we make those lints a hard error and unify the unsafe logic between `const fn` and `fn`.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [inline src/librustc_mir/transform/check_unsafety.rs]: Why only for mutable borrows?

What about `NonNull<Cell<i32>>`, and then taking an `&Cell<i32>` and writing `0` into it?

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [reply src/librustc_mir/transform/check_unsafety.rs]: > This all will clean up nicely when we make those lints a hard error and unify the unsafe logic between const fn and fn.

That is good.

File: src/test/ui/unsafe/ranged_ints.stderr
Comment (Centril) [reply src/test/ui/unsafe/ranged_ints.stderr]: Heh; we are making good error messages that only ever applies to compiler internals?
Wow, that's some dedication =P

Comment (oli-obk) [reply src/test/ui/unsafe/ranged_ints.stderr]: I can dump the valid value ranges, too, if you want. Miri already has code for that, we can put that code into `rustc_mir::utils` and share it.

Anything that makes developing the compiler easier is a good thing.

File: src/libcore/nonzero.rs
Comment (eddyb) [inline src/libcore/nonzero.rs]: Glad this is not exported anymore!

File: src/librustc_mir/transform/check_unsafety.rs
Comment (RalfJung) [inline src/librustc_mir/transform/check_unsafety.rs]: Looking good. :)  A few lines of comments would be helpful though, explaining in human readable form what is being checked here.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: oops, fixed

File: src/librustc_mir/transform/check_unsafety.rs
Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: done

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr
Comment (oli-obk) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr]: Doing that would require a little bit of spaghetti code just for the feature gate suggestion which would go away again with the stabilization of the feature

Comment (Centril) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr]: Fair -- hopefully it doesn't stay in nightly for long :)

Comment (oli-obk) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr]: Oh sorry, I forgot to mention. I found a non-spaghetti way :laughing: The code is actually cleaner now imo

Comment (Centril) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe.stderr]: Cool! ❤️

**Comment by oli-obk** [comment]
We do now

**Comment by Centril** [comment]
@RalfJung @oli-obk besides the rebase, what remains to do here?

**Comment by oli-obk** [comment]
rebased

> what remains to do here?

just needs another review I guess?

**Comment by Centril** [comment]
Assigning to Niko for final review.

r? @nikomatsakis

File: src/librustc/dep_graph/graph.rs
Comment (RalfJung) [inline src/librustc/dep_graph/graph.rs]: Why are new imports needed here now? This looks like an odd change.

File: src/libcore/ptr.rs
Comment (RalfJung) [inline src/libcore/ptr.rs]: Might be a bit nicer to use `new_unchecked`? (Same in the `From` instance below)

File: src/libsyntax/feature_gate.rs
Comment (RalfJung) [inline src/libsyntax/feature_gate.rs]: This comment does not match reality: You currently allow constructing layout-restricted types even without the feature flag.

I'd prefer if we did what the comment said, but at least they should be consistent.

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (RalfJung) [inline src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: Do we have a test making sure this is not even allowed with an unsafe block?

File: src/librustc/dep_graph/graph.rs
Comment (oli-obk) [reply src/librustc/dep_graph/graph.rs]: This is due to the `newtype_index` changes

Comment (eddyb) [reply src/librustc/dep_graph/graph.rs]: Can't `newtype_index` be written in a such a way that this isn't needed?

Comment (oli-obk) [reply src/librustc/dep_graph/graph.rs]: Some crates import `rustc_serialize` as `rustc_serialize` and some import it as `serialize`. So I can't just use global paths.

I also can't add my own `extern crate rustc_serialize;`... unless I put the trait impls into e.g. the `new` method where it won't conflict with other things in scope

File: src/libsyntax/feature_gate.rs
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: I don't think we can do what the comment says due to stability, sadly ;)

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (oli-obk) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: We do now

File: src/libsyntax/feature_gate.rs
Comment (oli-obk) [reply src/libsyntax/feature_gate.rs]: I changed the comment

File: src/libcore/ptr.rs
Comment (oli-obk) [reply src/libcore/ptr.rs]: done

File: src/librustc/dep_graph/graph.rs
Comment (oli-obk) [reply src/librustc/dep_graph/graph.rs]: done, that worked

File: src/librustc_data_structures/indexed_vec.rs
Comment (eddyb) [inline src/librustc_data_structures/indexed_vec.rs]: Heh, feel free to add a comment about using `const _ =` instead or something.
But also, why not just use absolute paths?
E.g. `impl ::serialize::Decodable for super::$type {` and `::serialize::Decoder::read_u32(d)`.

Comment (eddyb) [reply src/librustc_data_structures/indexed_vec.rs]: If you use `decl_macro` you could even rely on hygiene, but idk how @nikomatsakis and @petrochenkov feel about using that in the compiler.

Comment (Centril) [reply src/librustc_data_structures/indexed_vec.rs]: Well... I wouldn't bank on `decl_macro` staying the same given that it has barely gone through an experimental RFC...

Comment (oli-obk) [reply src/librustc_data_structures/indexed_vec.rs]: absolute paths don't work, because not all crates have `extern crate serialize;`. Some have `extern crate serialize as rustc_serialize;`

File: src/libsyntax/feature_gate.rs
Comment (RalfJung) [reply src/libsyntax/feature_gate.rs]: > I don't think we can do what the comment says due to stability, sadly ;)

Why not? Just enable this `min_unsafe_const_fn` feature in libstd?

Comment (oli-obk) [reply src/libsyntax/feature_gate.rs]: I think @Centril is referring to the fact that you can't use unstable const features in `min_const_fn` functions. But these unsafety checks are an exception.

That said, I think there's little point to this, as these are technically already behind the `rustc_attrs` feature gate, as you can't define such layout restricted types without unstable attributes. They can only really be used in the compiler or other unstable code. Having to enable another feature gate doesn't protect anyone from using unstable features.

File: src/librustc_data_structures/indexed_vec.rs
Comment (eddyb) [reply src/librustc_data_structures/indexed_vec.rs]: Right, I was hoping that wouldn't be a problem but... we'll have to wait for serde/some rustc-specific custom serialization solution, post-#49219.

**Comment by oli-obk** [comment]
This PR should be ready for a final review now

File: src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs
Comment (nikomatsakis) [reply src/test/ui/consts/min_const_fn/min_const_fn_unsafe_feature_gate.rs]: It may indeed be worth having an extra note like:

```
= note: unsafe actions within a `const unsafe fn` still require an `unsafe` block
```

**Review by nikomatsakis — APPROVED** [review]
Looks nice! I left a bunch of nits. r=me once they are fixed, I think.

**Comment by nikomatsakis** [comment]
To be clear, where I wrote:

> It may indeed be worth having an extra note like:

I'm ok with deferring this till later, but I think it'll be a source of confusion.

File: src/librustc/ty/constness.rs
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: That doesn't work, because the intrinsics do not have "const" in their signature. ... Which makes me wonder why not. I'll see if I can get intrinsics to be marked as const at their definition, so we can simplify all this.

Comment (oli-obk) [reply src/librustc/ty/constness.rs]: I'm gonna start a separate PR for that and rebase this PR over the new one.

File: src/librustc_mir/transform/check_unsafety.rs
Comment (oli-obk) [reply src/librustc_mir/transform/check_unsafety.rs]: I seem to have forgotten to push some cleanups where this is just `bug!`

**Comment by oli-obk** [comment]
Uh... @alexcrichton @nrc what is the correct procedure for landing something that breaks a `rustc-ap-*` crate needed for cargo?

**Comment by alexcrichton** [comment]
@oli-obk in theory there's no procedure as we just publish a new version of rustc-ap-syntax every night, whatever the nightly happened to be released with. I've just fixed a bug where it stopped running for a week or so, so maybe that's all that was needed?

Does that help? I fear I may be missing context about what's actually needed here

**Comment by oli-obk** [comment]
I made a breaking change to `librustc_data_structures`, which in turn broke `rustc-ap-syntax_pos`.

From your statement it sounds like that's normally fine, so it might be that that in turn would have broken rls, and since we're in publishing week, we can't break tools.

I am currently investigating if I can work around it by a targeted application of `#[allow_internal_unstable]`.

**Comment by alexcrichton** [comment]
Hm that's odd because that breakage should be impossible, the `rustc-ap-syntax_pos` crate depend on `rustc-ap-rustc_data_structures`, not the in-tree copy. Is the breakage in the logs above or elsewhere?

**Comment by oli-obk** [comment]
> Is the breakage in the logs above or elsewhere?

It's in the logs above.... But I think I know the issue: I changed language semantics for the forever-unstable feature `rustc_layout_scalar_valid_range` attributes. Using them requires `unsafe` now.

**Comment by alexcrichton** [comment]
Ah yes indeed! If that breaks crates on crates.io that's fine, it just means that until the release happens (where we require the RLS compiles) this won't be able to land. After the release though this can land, it'll break the RLS, and then the RLS will update to a newer version of the crate on crates.io as it's naturally published.

File: src/libcore/num/mod.rs
Comment (pnkfelix) [inline src/libcore/num/mod.rs]: Maybe I missed something: why is the inner `unsafe` necessary here within the body of an `unsafe fn` ? Does const-eval cause an effect analogous to https://github.com/rust-lang/rfcs/pull/2585 ? Because I am seeing warnings about an unnecessary `unsafe` when I do bootstrap builds right now.

Seems like it would warrant at least a comment in the code...

Comment (oli-obk) [reply src/libcore/num/mod.rs]: That [was discussed](https://github.com/rust-lang/rust/issues/55607#issuecomment-435549534) on the tracking issue, and [recently discussed to not do anymore](https://github.com/rust-lang/rust/issues/55607#issuecomment-445882296). The revert PR is https://github.com/rust-lang/rust/pull/56706

Comment (pnkfelix) [reply src/libcore/num/mod.rs]: Ah okay thanks

**Pull request merged**
